samples/nrf9160/serial_lte_modem/src/slm_at_host.c

```diff
@@ -59,19 +67,40 @@ static size_t at_buf_len;
 static struct k_work cmd_send_work;
 static const char termination[3] = { '\0', '\r', '\n' };
 
+static u8_t uart_rx_buf[UART_RX_LEN][1];
+static char *uart_tx_buf;
+static u8_t buf_num;
+
+K_SEM_DEFINE(tx_done, 0, 1);
+
+/* global functions defined in different files */
+void enter_idle(void);
+void enter_sleep(void);
+
 /* global variable defined in different files */
 extern struct at_param_list m_param_list;
-extern void enter_sleep(void);
 
 /* forward declaration */
 void slm_at_host_uninit(void);
 
-static inline void write_uart_string(const char *str, size_t len)
+static void write_uart_string(const char *str, size_t len)
 {
+	int ret;
+
+	k_sem_take(&tx_done, K_FOREVER);
+
+	uart_tx_buf = k_malloc(len);
```

Comment:
Do we really need dynamic allocation here? Wouldn't it be simpler to preallocate a reasonable buffer for this purpose?

Reply:
Updated to use a static pointer to hold the RAM buffer for DMA purpose.